### DISCUSSION TITLE
Request for support ARCO ERA5 data integration

### DISCUSSION THREAD

**Discussion by P-Myers** [WAM2layers/WAM2layers#415] [Q&A] [answered]
Hi everyone,

As per #401, in which script has been added to integrate the ARCO online hosted ERA5 data into the model on the fly, I wanted to address the following:

1.) Are there any recommendations on how to trim the ARCO data to 75, -75 lat to improve calculation performance? When I run the preprocess, some NaNs show up (in my case, for precip). I believe this may be connected to the lat boundaries of the ARCO data called in the script, which upon inspection of the data are 90,-90. Consequently, a broadcasting error appeared when tracking was conducted, as the base grid of my tagging shapefile.nc is trimmed to 75, -75 lat. 

I had initially attempted to trim the ARCO data to improve model performance by setting the tracking domain to [-180, -75, 180, 75], which resulted in an 601 lat points and an unexpected 721 lon points in the fluxes_storages file. The ARCO data uses the range 0 - 359.75 for its longitude, making the use of a tracking domain buggy as the model's config.py file is set to work with -180, 180. I imagine the 721 was produced as a result of an error with how the script handles the native 0, 360 lon grid of the ARCO data. I proceeded to adjust the expectations of the model to account for this, but did not succeed. 

2.) Can the script be adapted to use model level data? Currently, it is only set up for pressure level use, and I am uncertain of what precise adjustments need to be made for this. 

Thanks!

Parker

**Comment by P-Myers** [comment]
I was able to trim the domain and produce preprocessed files that reflected this, however the presence of NaNs remained at the date I was encountering them. 

With some checking, I was able to pinpoint the source to be the raw data 'surface_latent_heat_flux.' Here is the error code with some print statements:

![image](https://github.com/user-attachments/assets/afa9f1a4-266b-4632-8b29-694c52635e74)

What is the best way to address this?

Any insight is much appreciated.

**Reply by BSchilperoort** [reply]
Hi Parker, I don't have time right now to finish up #401 myself, but I can try to point you the right way. If you end up fixing issues with the code, could you make a pull request for your branch so the fixes can be merged into WAM2layers? The documentation has[ a page on how to help with development](https://wam2layers.readthedocs.io/en/latest/develop.html).

It's strange that there are NaN values in the data. But it could be that this has to do with the original data source ("raw" ARCO-ERA5 data). Seeing as you know the time step and variable you could explore this. See the [ARCO-ERA5 repository](https://github.com/google-research/arco-era5?tab=readme-ov-file#025-pressure-and-surface-level-data) for how to open the dataset yourself outside of WAM2layers.

What can also help in debugging is either:
1. Use IPython (`from IPython import embed; embed(); quit();`) to allow you to inspect the variables at some point in the code when you use the CLI.
2. Call the preprocessing code in a .py file and use the python debugger in your IDE

**Reply by P-Myers** [reply]
Hi Bart,

I had some success with a simple nearest neighbor interpolation applied in the arco.py preprocess code. NaNs seem to be throughout the surface_latent_heat_flux variable in the arco set. 

Unfortunately this success was short lived, as I am now experiencing a netCDF append issue. I have redownloaded the branch to see if it was related to changes I had made, but the issue persists. Working on debugging this now, any tips would be greatly appreciated! 

![image](https://github.com/user-attachments/assets/16558b07-53de-4831-9688-06d2b1f6d5b2)

Perhaps its something in my config.yaml?

![image](https://github.com/user-attachments/assets/cc731e5c-8c40-4555-92f8-bfb0ac04f4eb)

**Comment by P-Myers** [comment]
Wanted to follow up on this thread, latest updates in #401 produce much more stable outputs, thanks for that! 

Could an example arco config file perhaps be added into the docs for reference? I believe my issues now are related to my setup.

Currently, I encounter the following error output when visualizing input data: TypeError: Cannot convert input [2017-06-01 00:00:00] of type <class 'cftime._cftime.DatetimeProlepticGregorian'> to Timestamp

When visualizing tracking, I get this error when output_frequency is "1d": ValueError: Invalid frequency string provided

Using "1D" works. However my tracking visualization produces a concentrated blob, and does not align with my ERA5 data from the Copernicus portal. 

Thank you!

**Reply by BSchilperoort** [reply, marked as answer]
Hi Parker, I did find some mistake I made in the ARCO preprocessor, to that's why your outputs are better (hopefully correct).

I have now validated it on preprocessed ERA5 data, and it gave the same result apart from floating point precision inaccuracies. The median deviation was about 10e-4 %.

> Currently, I encounter the following error output when visualizing input data: TypeError: Cannot convert input [2017-06-01 00:00:00] of type <class 'cftime._cftime.DatetimeProlepticGregorian'> to Timestamp

That would not be related to something you do wrong. We had to change to CF timestamps to support calendars that climate models use, but it seems I still have to update the visualization routines.



> When visualizing tracking, I get this error when output_frequency is "1d": ValueError: Invalid frequency string provided
> 
> Using "1D" works. However my tracking visualization produces a concentrated blob, and does not align with my ERA5 data from the Copernicus portal.

Can you open an issue on this? I think the main problem is that pandas changed in what a valid frequency string is ("1D", "6h" are valid, "1d", "6H" are not). See [their docs](https://pandas.pydata.org/docs/reference/api/pandas.Timedelta.html#pandas-timedelta) for the valid arguments. However, this is not clearly documented, and the example config on the repository does use "1d".
